modules/core/keeper/msg_server.go

```diff
@@ -494,6 +494,9 @@ func (k Keeper) RecvPacket(goCtx context.Context, msg *channeltypes.MsgRecvPacke
 	if ack == nil || ack.Success() {
 		// write application state changes for asynchronous and successful acknowledgements
 		writeFn()
+	} else {
+		// Modify events in cached context to reflect unsuccessful acknowledgement
+		ctx.EventManager().EmitEvents(coretypes.ConvertToErrorEvents(cacheCtx.EventManager().Events()))
```

Comment:
```suggestion
	} else if !ack.Success() {
		// Modify events in cached context to reflect unsuccessful acknowledgement
		ctx.EventManager().EmitEvents(coretypes.ConvertToErrorEvents(cacheCtx.EventManager().Events()))
```
I don't think the intention is to emit error events for async acks? Might be best to be explicit?

Reply:
I think this suggestion is already enforced. It is not possible for `ack.Success()` to be true here anyway. Async acknowledgements are handled in the [if case](https://github.com/cosmos/ibc-go/blob/07fca7efc918dbc8e6cc04f13fd04c101bb87664/modules/core/keeper/msg_server.go#L494). Is there some misunderstanding?

Reply:
I think async acks are handled in `ack == nil` but I don't see the harm in being explicit with `else if !ack.Success()`

Reply:
ah yes, that's correct. I misread the first if statement